src/msg/async/AsyncMessenger.cc

```diff
@@ -601,34 +601,42 @@ AsyncConnectionRef AsyncMessenger::create_connect(
   return conn;
 }
 
-ConnectionRef AsyncMessenger::connect_to(int type, const entity_addrvec_t& addrs)
+
+ConnectionRef AsyncMessenger::get_loopback_connection()
 {
-  Mutex::Locker l(lock);
-  if (*my_addrs == addrs ||
-      (addrs.v.size() == 1 &&
-       my_addrs->contains(addrs.front()))) {
-    // local
-    return local_connection;
-  }
+  return local_connection;
+}
 
-  AsyncConnectionRef conn = _lookup_conn(addrs);
-  if (conn) {
-    ldout(cct, 10) << __func__ << " " << addrs << " existing " << conn << dendl;
+entity_addrvec_t AsyncMessenger::_filter_addrs(int type,
+					       const entity_addrvec_t& addrs)
+{
+  if (did_bind &&
+      !get_myaddrs().has_msgr2() &&
+      get_mytype() == type) {
```

Comment:
why does filtering is only applied when the peer type is equal to this messenger type?

Reply:
that is the only time there is a stateful, bidirection connection: mon <-> mon, osd <-> osd, mds <-> mds.  The rest of the time there is a clear client and server, the client is responsible for reconnecting, so the protocol version asymmetry isn't a problem.

Reply:
gotcha.

Reply:
Hmm, @batrick , we might need to do something with the client vs mds connections here.  Are the clients identified only by entity_name_t or also by entity_addr_t?  Becuase the addr may vary between MDSs for the same client is some MDSs have restarted with a v2 addr and some haven't.

RBD had a similar problem and we "fixed" it by normalizing on a v1 addr always in the cls_lock code.

Reply:
Looks like it uses just `entity_name_t`: https://github.com/ceph/ceph/blob/b57f896ed5333502ec3147cba0ea5340e1e1b46f/src/mds/MDSDaemon.cc#L1311

So this is not a problem?

Reply:
yeah it looks good.  thanks!